### PR TITLE
Fix generic module interface

### PR DESCRIPTION
**Pull request by ntunin**
## Description
Improved generic module architecture, public properties guarded by according protocols

## Related tasks
[ISD-667](https://rosberry.atlassian.net/browse/ISD-667)

## TODOS:
- Update ReadME

### CODE REVIEW

File: Sources/GenericModule/Presenter/Presenter.swift [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+    open func viewWillDisappear() {
+
+    }
+
+    open func viewDidDisappear() {
+
+    }
+
+    open var dependencies: Dependencies
```
Comment (zhvrnkov) [inline Sources/GenericModule/Presenter/Presenter.swift:41]: Make as const and move up to properties?

File: Sources/GenericModule/Presenter/Presenter.swift [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+                     Output,
+                     Dependencies>: ModulePresenter, ViewOutput {
+
+    public typealias ViewModel = View.ViewModel
+    public typealias State = ViewModel.State
+
+    public var state: State
+    public var output: Output?
+
+    weak var view: View?
```
Comment (zhvrnkov) [inline Sources/GenericModule/Presenter/Presenter.swift:16]: I wonder if we can make it as unowned and view become not optional

File: Sources/GenericModule/View/View.swift [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+//
+
+public protocol View: class {
+    associatedtype Output
+    associatedtype ViewModel: GenericModule.ViewModel
+    associatedtype ViewInput
+
+    var output: Output? { get set }
+
+    init(viewModel: ViewModel)
```
Comment (zhvrnkov) [inline Sources/GenericModule/View/View.swift:12]: What do you think about passing Output to init and View.output become not optional? Is it possible?

**Review by Tublad — APPROVED** [review]

File: Sources/GenericModule/Presenter/Presenter.swift (hunk 2 above)
Comment (ntunin) [reply Sources/GenericModule/Presenter/Presenter.swift:17]: view input is not optional now

File: Sources/GenericModule/View/View.swift (hunk 3 above)
Comment (ntunin) [reply Sources/GenericModule/View/View.swift]: view output is not optional now

**Review by zhvrnkov — APPROVED** [review]

**Review by disabled — APPROVED** [review]
Collection view insets are not accounted for when cell size is calculated at `SecondViewController.swift:122`, which causes it to throw a warning to the log.
Otherwise, looks solid.

**Pull request merged**